src/dotnet/Chat.Service/Authors.cs

```diff
@@ -228,8 +230,60 @@ public virtual async Task OnInvite(Authors_Invite command, CancellationToken can
         var chat = await Chats.Get(session, chatId, cancellationToken).Require().ConfigureAwait(false);
         chat.Rules.Require(ChatPermissions.Invite);
 
-        foreach (var userId in userIds)
-            await Backend.EnsureJoined(chatId, userId, cancellationToken).ConfigureAwait(false);
+        foreach (var userId in userIds) {
+            var author = await Backend.EnsureJoined(chatId, userId, cancellationToken).ConfigureAwait(false);
+            if (author.HasLeft)
+                await RestoreAuthorMembership(cancellationToken, author).ConfigureAwait(false);
+        }
+    }
+
+    // [CommandHandler]
+    public virtual async Task OnExclude(Authors_Exclude command, CancellationToken cancellationToken)
+    {
+        if (Computed.IsInvalidating())
+            return; // It just spawns other commands, so nothing to do here
+
+        var (session, authorId) = command;
+        var chatId = authorId.ChatId;
+        var chat = await Chats.Get(session, chatId, cancellationToken).Require().ConfigureAwait(false);
+        chat.Rules.Require();
+        chat.Rules.Require(ChatPermissions.EditMembers);
+
+        var author = await Backend.Get(chatId, authorId, cancellationToken).ConfigureAwait(false);
+        if (author == null || author.HasLeft)
+            return;
+
+        if (chat.Rules.Account.Id == author.UserId)
+            throw StandardError.Constraint("You can not remove yourself from the chat members.");
+
+        var ownerIds = await Roles.ListOwnerIds(session, chatId, cancellationToken).ConfigureAwait(false);
+        var isOwner = ownerIds.Contains(authorId);
+        if (isOwner)
+            throw StandardError.Constraint("You can not remove an owner from the chat members.");
+
+        var upsertCommand = new AuthorsBackend_Upsert(
+            chatId, author.Id, default, author.Version,
+            new AuthorDiff() { HasLeft = true });
+        await Commander.Call(upsertCommand, true, cancellationToken).ConfigureAwait(false);
+    }
+
+    // [CommandHandler]
+    public virtual async Task OnRestore(Authors_Restore command, CancellationToken cancellationToken)
+    {
+        if (Computed.IsInvalidating())
+            return; // It just spawns other commands, so nothing to do here
+
+        var (session, authorId) = command;
+        var chatId = authorId.ChatId;
+        var chat = await Chats.Get(session, chatId, cancellationToken).Require().ConfigureAwait(false);
+        chat.Rules.Require();
+        chat.Rules.Require(ChatPermissions.EditMembers);
```

Comment:
assertion code seems to be same for OnExclude and OnRestore

Reply:
Removed `chat.Rules.Require();` since it's redundant.